### PR TITLE
Add target peer count range

### PR DESCRIPTION
**Pull request by ajsutton**
## PR Description
`ConnectionManager` now maintains a target range for number of peers to be connected to.  New peers are only connected when the current peer count is below the target range, with enough connection attempts made to reach the top of that range, expecting that some will fail to connect.

When peers subscribe that push us above the target range, we disconnect some to get back down to the range.  We don't currently track peer reputation so an arbitrary peer is picked to disconnect.

### CODE REVIEW

File: networking/p2p/src/test/java/tech/pegasys/artemis/networking/p2p/discovery/ConnectionManagerTest.java [hunk 1]
```diff
@@ -251,7 +256,92 @@ public void shouldConnectToKnownPeersWhenDiscoverySearchCompletes() {
     verify(network).connect(discoveryPeer2);
   }
 
+  @Test
+  public void shouldLimitNumberOfNewConnectionsToKnownPeersOnStartup() {
```
Comment (macfarla) [inline networking/p2p/src/test/java/tech/pegasys/artemis/networking/p2p/discovery/ConnectionManagerTest.java]:
```suggestion
  public void shouldLimitNumberOfNewConnectionsToRangeUpperBoundOnStartup() {
```

Comment (macfarla) [reply networking/p2p/src/test/java/tech/pegasys/artemis/networking/p2p/discovery/ConnectionManagerTest.java]: nit - on first read I took it as limiting the connections to only known peers

File: networking/p2p/src/test/java/tech/pegasys/artemis/networking/p2p/discovery/ConnectionManagerTest.java [hunk 2]
```diff
@@ -251,7 +256,92 @@ public void shouldConnectToKnownPeersWhenDiscoverySearchCompletes() {
...
+
+    manager.start().join();
+
+    verify(network).connect(discoveryPeer1);
+    verify(network).connect(discoveryPeer2);
+    verify(network, never()).connect(discoveryPeer3);
+  }
+
+  @Test
+  public void shouldLimitNumberOfNewConnectionsToKnownPeersOnRetry() {
```
Comment (macfarla) [inline networking/p2p/src/test/java/tech/pegasys/artemis/networking/p2p/discovery/ConnectionManagerTest.java]:
```suggestion
  public void shouldLimitNumberOfNewConnectionsToRangeUpperBoundOnRetry() {
```

**Review by macfarla — APPROVED** [review]
LGTM. suggested changing some test names

File: networking/p2p/src/test/java/tech/pegasys/artemis/networking/p2p/discovery/ConnectionManagerTest.java (hunk 1 above)
Comment (ajsutton) [reply networking/p2p/src/test/java/tech/pegasys/artemis/networking/p2p/discovery/ConnectionManagerTest.java]: We aren't limiting the total number of connections, just the one to discovery peers.  Have clarified along those lines.

**Pull request merged**